libs/cpp-ledger/ledger/apdu_wrapper.cpp

```diff
@@ -0,0 +1,198 @@
+    }
+
+    return std::make_tuple(result, offset, Error{});
+  }
+
+  std::tuple<Bytes, uint16_t, Error> DeserializePacket(uint16_t channel,
+                                                       const Bytes &buffer,
+                                                       uint16_t sequenceId) {
+    if ((sequenceId == 0 && buffer.size() < 7)
+        || (sequenceId > 0 && buffer.size() < 5)) {
+      return std::make_tuple(
+          Bytes{},
+          0,
+          Error{
+              "Cannot deserialize the packet. Header information is missing."});
+    }
+
+    Byte headerOffset = 0;
+
+    if (getFromBytes(buffer[0], buffer[1]) != channel) {
+      return std::make_tuple(Bytes{}, 0, Error{"Invalid channel"});
+    }
+    headerOffset += 2;
+
+    if (buffer[headerOffset] != 0x05) {
+      return std::make_tuple(Bytes{}, 0, Error{"Invalid tag"});
+    }
+    headerOffset++;
+
+    if (getFromBytes(buffer[headerOffset], buffer[headerOffset + 1])
+        != channel) {
+      return std::make_tuple(Bytes{}, 0, Error{"Wrong sequenceId"});
+    }
+    headerOffset += 2;
+
+    uint16_t totalResponseLength = 0;
+
+    if (sequenceId == 0) {
+      totalResponseLength =
+          getFromBytes(buffer[headerOffset], buffer[headerOffset + 1]);
+      headerOffset += 2;
+    }
+
+    const Bytes result(buffer.begin() + headerOffset, buffer.end());
+
+    return std::make_tuple(result, totalResponseLength, Error{});
+  }
+
+  std::tuple<Bytes, Error> WrapCommandApdu(uint16_t channel,
+                                           Bytes command,
+                                           int packetSize) {
+    Bytes totalResult;
+    uint16_t sequenceId = 0;
+    while (!command.empty()) {
+      const auto [result, offset, err] =
+          SerializePacket(channel, command, packetSize, sequenceId);
+      if (err != std::nullopt) {
+        return std::make_tuple(Bytes{}, err);
+      }
+      command = Bytes(command.begin() + offset, command.end());
+      totalResult.insert(totalResult.end(), result.begin(), result.end());
+      sequenceId++;
+    }
+
+    return std::make_tuple(totalResult, Error{});
+  }
+
+  std::tuple<Bytes, Error> UnwrapResponseApdu(uint16_t channel,
+                                              const std::vector<Bytes> &packets,
+                                              int packetSize) {
+    Bytes totalResult;
+    uint16_t totalSize = 0;
+    uint16_t sequenceId = 0;
+
+    for (const auto &packet : packets) {
+      const auto [result, responseSize, err] =
+          DeserializePacket(kChannel, packet, sequenceId);
+
+      if (err != std::nullopt) {
+        return std::make_tuple(Bytes{}, err);
+      }
+
+      if (sequenceId == 0) {
+        totalSize = responseSize;
+        totalResult.reserve(totalSize);
+      }
+
+      totalResult.insert(totalResult.end(), result.begin(), result.end());
+      sequenceId++;
+    }
+
+    totalResult = Bytes(totalResult.begin(), totalResult.begin() + totalSize);
```

Comment:
maybe return error if packets make less/greater (not equal) bytes?  
out-of-bound vector access.